ros/fsai_mission/scripts/lap_counter.py

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+
+# This node counts the laps by drawing a line connecting large orange cones and detecting cross product sign change
+
+import rospy
+from geometry_msgs.msg import Point
+from ultralytics_ros.msg import ConeDetection
+from nav_msgs.msg import Odometry
+from std_msgs.msg import UInt16
+import math
+
+class LapCounter:
+    def __init__(self):
+        rospy.init_node('lap_counter', anonymous=False)
+
+        # Subscribers
+        self.cone_sub = rospy.Subscriber('/cone_list', ConeDetection, self.cone_callback)
+        self.odom_sub = rospy.Subscriber('/odom', Odometry, self.odom_callback)
+
+        # Publisher
+        self.lap_counter_pub = rospy.Publisher('/laps', UInt16, queue_size=10)
+
+        # State
+        self.car_position = Point()
+        self.cone_positions = []
+        self.previous_position = None
+        self.lap_count = 0  # Lap count
+        self.in_cooldown = False
+
+        # Timer for periodic publishing
+        self.publish_timer = rospy.Timer(rospy.Duration(1), self.publish_lap_count)
+
+        rospy.loginfo("Lap Counter node initialised")
+
+    def cone_callback(self, msg):
+        self.cone_positions = msg.large_orange_cones
+
+    def odom_callback(self, msg):
+        self.car_position = msg.pose.pose.position
+
+        if self.previous_position is not None and len(self.cone_positions) >= 2 and not self.in_cooldown:
+            first_cone = self.cone_positions[0]
+            second_cone = self.cone_positions[1]
+
+            # Calculate vectors
+            finish_line_vector = (second_cone.x - first_cone.x, second_cone.y - first_cone.y)
+            prev_car_vector = (self.previous_position.x - first_cone.x, self.previous_position.y - first_cone.y)
+            curr_car_vector = (self.car_position.x - first_cone.x, self.car_position.y - first_cone.y)
+
+            # Calculate cross products to determine if the car has crossed the finish line
+            prev_cross = finish_line_vector[0] * prev_car_vector[1] - finish_line_vector[1] * prev_car_vector[0]
+            curr_cross = finish_line_vector[0] * curr_car_vector[1] - finish_line_vector[1] * curr_car_vector[0]
```

Comment:
Your can use `np.cross`

Reply:
I can, but why are we trying to fix something that is not broken? This method is also way more efficient than using numpy which uses overheads for function calls

Reply:
It's cleaner if you use np.cross, and I'd give it 80% chance that it would be faster.
If we don't have time to write nice code, I'll stop adding comments like that

Reply:
You're right, it's faster without np.